src/main.ts

```diff
@@ -0,0 +1,258 @@
+ */
+async function parseInputs({
+  argv,
+  outputLogger,
+  defaultProjectNames,
+}: {
+  argv: string[];
+  outputLogger: OutputLogger;
+  defaultProjectNames: string[];
+}): Promise<Inputs | null> {
+  const args = await parseCommandLineArguments({ argv, outputLogger });
+
+  if (args) {
+    const projectReferences = args._.length > 0 ? args._ : defaultProjectNames;
+    return {
+      templateRepositoryName: DEFAULT_TEMPLATE_REPOSITORY_NAME,
+      projectReferences,
+    };
+  }
+
+  /* istanbul ignore next: At the moment, there is no real way that Yargs could fail */
+  return null;
+}
+
+/**
+ * Parses the command line using `yargs`.
+ *
+ * @param args - The arguments to this function.
+ * @param args.argv - The name of this executable and its arguments (as obtained
+ * via `process.argv`).
+ * @param args.outputLogger - Writable streams for output messages.
+ */
+async function parseCommandLineArguments({
+  argv,
+  outputLogger,
+}: {
+  argv: string[];
+  outputLogger: OutputLogger;
+}) {
+  let yargsFailure: { message: string; error: Error } | null = null;
+  const onFail = (message: string, error: Error) => {
+    /* istanbul ignore next: Not sure how to produce this */
+    if (error) {
+      throw error;
+    }
+    yargsFailure = { message, error };
+  };
+
+  const parser = yargs(hideBin(argv))
+    .usage(USAGE_TEXT)
+    .help(false)
+    .version(false)
+    .option('help', {
+      alias: 'h',
+      describe: 'Show help',
+      type: 'boolean',
+    })
+    .wrap(null)
+    .exitProcess(false)
+    .fail(onFail)
+    .strictOptions();
+
+  const options = await parser.parse();
+
+  if (options.help) {
+    outputLogger.logToStderr(await parser.getHelp());
+    return null;
+  }
+
+  if (yargsFailure) {
+    const { message } = yargsFailure;
+    outputLogger.logToStderr(
+      'ERROR: %s\n\n%s',
+      message,
+      await parser.getHelp(),
+    );
+    return null;
+  }
+
+  return {
+    ...options,
+    // This is the name that Yargs gives to the arguments that aren't options.
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    _: options._.map(String),
+  };
+}
+
+/**
+ * Runs all of the rules on the given projects.
+ *
+ * This is a bit complicated because linting may fail for a particular project.
+ * In that case, we don't want to crash the whole tool, but fail gracefully.
+ *
+ * @param args - The arguments to this function.
+ * @param args.projectReferences - References to the projects (either the name
+ * of a MetaMask repository, such as "utils", or the path to a local Git
+ * repository).
+ * @param args.template - The repository to which the project should be
+ * compared.
+ * @param args.workingDirectoryPath - The directory where this tool was run.
+ * @param args.cachedRepositoriesDirectoryPath - The directory where MetaMask
+ * repositories will be (or have been) cloned.
+ * @param args.outputLogger - Writable streams for output messages.
+ */
+async function lintProjects({
```

Comment:
Great idea to use `allSettled` here to handle failures gracefully, and to put failures last. Seems good.

Thoughts on adding a little summary at the end? e.g X/Y passed

Reply:
Good idea, added here: [8a025a0](https://github.com/MetaMask/module-lint/pull/33/commits/8a025a06b5978bea7f13a295f6c54337af5719bf)

Reply:
Nice!